File: server/internal/domain/sessions/sessions.go
package sessionsdomain

import (
	"time"

	sessionspb "github.com/aesterial/secureguard/internal/api/v1/sessions/v1"
	"google.golang.org/protobuf/types/known/timestamppb"
)

type Session struct {
	ID        string
	Hash      string
	Revoked   bool
	RevokedAt *time.Time
	Created   time.Time
	Expires   time.Time
	LastSeen  *time.Time
}

func (s *Session) rvProtobuf() *sessionspb.Session_RevokeInfo {
	var revokedAt *timestamppb.Timestamp = nil
	if s.Revoked {
		revokedAt = timestamppb.New(*s.RevokedAt)
	}
	return &sessionspb.Session_RevokeInfo{
		Revoked:   s.Revoked,
		RevokedAt: revokedAt,
	}
}

func (s *Session) Protobuf() *sessionspb.Session {
	if s == nil {
		return nil
	}
	var lastSeen *timestamppb.Timestamp = nil
	if s.LastSeen != nil {
		lastSeen = timestamppb.New(*s.LastSeen)
	}
	return &sessionspb.Session{
		Id:        s.ID,
		Hash:      s.Hash,
		Revoke:    s.rvProtobuf(),
		CreatedAt: timestamppb.New(s.Created),
		ExpiresAt: timestamppb.New(s.Expires),
		LastSeen:  lastSeen,
	}
}

type Sessions []*Session

func (s *Sessions) Protobuf(skipRevoked ...bool) []*sessionspb.Session {
	if s == nil {
		return nil
	}
	var skipRv bool
	if len(skipRevoked) > 0 {
		skipRv = skipRevoked[0]
	}
	var list = make([]*sessionspb.Session, len(*s))
	for i, e := range *s {
		if e.Revoked && skipRv {
			continue
		}
		list[i] = e.Protobuf()
	}
	return list
}

File: client/cli/lib/src/models/sessions.dart
import 'package:secureguard_cli/src/api/xyz/secureguard/v1/sessions/v1/domain.pb.dart'
    as $session;

class Session {
  final String id;
  final String hash;
  final RevokeInfo? revoke;
  final DateTime? created;
  final DateTime? expires;
  final DateTime? lastSeen;
  const Session({
    required this.id,
    required this.hash,
    required this.revoke,
    required this.created,
    required this.expires,
    required this.lastSeen,
  });

  static Session? tryFromProto($session.Session session) {
    if (!session.hasId() || !session.hasHash()) {
      return null;
    }

    return Session(
      id: session.id,
      hash: session.hash,
      revoke: session.hasRevoke() ? RevokeInfo.fromProto(session.revoke) : null,
      created: session.hasCreatedAt() ? session.createdAt.toDateTime() : null,
      expires: session.hasExpiresAt() ? session.expiresAt.toDateTime() : null,
      lastSeen: session.hasLastSeen() ? session.lastSeen.toDateTime() : null,
    );
  }

  factory Session.fromProto($session.Session session) {
    final mapped = tryFromProto(session);
    if (mapped == null) {
      throw StateError('Session payload is missing required identity fields');
    }
    return mapped;
  }
}

class RevokeInfo {
  final bool revoked;
  final DateTime? at;
  const RevokeInfo({required this.revoked, required this.at});

  factory RevokeInfo.fromProto($session.Session_RevokeInfo revokeInfo) {
    return RevokeInfo(
      revoked: revokeInfo.revoked,
      at: revokeInfo.hasRevokedAt() ? revokeInfo.revokedAt.toDateTime() : null,
    );
  }
}
